ballista/rust/scheduler/src/scheduler_server/event_loop.rs

```diff
@@ -107,14 +125,19 @@ impl<T: 'static + AsLogicalPlan, U: 'static + AsExecutionPlan>
                             }
                         })
                         .collect::<Vec<String>>(),
-                    executor_data.executor_id
+                    executor_delta_data.executor_id
                 );
                 let mut client = {
                     let clients = self.executors_client.read().await;
-                    clients.get(&executor_data.executor_id).unwrap().clone()
+                    clients
+                        .get(&executor_delta_data.executor_id)
+                        .unwrap()
+                        .clone()
                 };
                 // Update the resources first
-                self.state.save_executor_data(executor_data.clone());
+                self.state
+                    .executor_manager
+                    .update_executor_data(executor_delta_data);
                 // TODO check whether launching task is successful or not
                 client.launch_task(LaunchTaskParams { task: tasks }).await?;
```

Comment:
Hi @thinkharderdev, here, I didn't consider too much about the error handling previously. I think you suggestion of doing it reverse makes sense.

Before updating the executor data, the tasks status have already been updated into Running status. Whether task launch successful or not, these tasks will be dangling in Running status. It's better for us to leverage speculative task execution in the future. And we don't need to consider too much of the situation of partial tasks launching success.

Why do you think we need a write lock?

Reply:
It was just a thought, not sure if the performance cost would be justifies. But the idea is that updating the executor state and launching the tasks should in principle be an atomic transaction. If you update first and then the launching fails, you resources will never be released but if you launch first and then update state then you have a concurrency issue where another thread may try and launch tasks after the task are launched and before the state is updated so you end up overcommitted. 

The second issue seems less of a problem because it seems pretty straightforward to just have the executor reject the task if it has no available task slots, so it may be enough to just reverse the order here (launch tasks -> update state) and let the existing error handling mechanisms deal with it. Since launching requires a network call, holding a write lock on the entire state for that operation could be a major performance bottleneck, so if we did need a lock it would probably have to be at the individual executor level.

Reply:
Thanks @thinkharderdev. Actually, for the current design, the method *launch_tasks* will only be invoked in the single consumer side of the channel, which means there will be no concurrency issue.

Reply:
However, as you mentioned, the executor available task slots in the scheduler side is just kind of guidance for task scheduling. The real source of truth is in the executor side.